analyzers/tests/SonarAnalyzer.Test/Rules/AspNet/ApiControllersShouldNotDeriveDirectlyFromControllerTest.cs

```diff
@@ -0,0 +1,116 @@
+
+#if NET
+    private static IEnumerable<MetadataReference> AspNetCoreReferences =>
+    [
+        AspNetCoreMetadataReference.MicrosoftAspNetCoreMvcAbstractions,
+        AspNetCoreMetadataReference.MicrosoftAspNetCoreMvcCore,
+        AspNetCoreMetadataReference.MicrosoftAspNetCoreMvcViewFeatures,
+    ];
+
+    [TestMethod]
+    public void ApiControllersShouldNotDeriveDirectlyFromController_CS() =>
+        builder.AddReferences(AspNetCoreReferences).AddPaths("ApiControllersShouldNotDeriveDirectlyFromController.cs").Verify();
+
+    [DataRow("""View()""")]
+    [DataRow("""View("viewName")""")]
+    [DataRow("""View(model)""")]
+    [DataRow("""View("viewName", model)""")]
+    [DataRow("""PartialView()""")]
+    [DataRow("""PartialView("viewName")""")]
+    [DataRow("""PartialView(model)""")]
+    [DataRow("""PartialView("viewName", model)""")]
+    [DataRow("""ViewComponent("foo")""")]
+    [DataRow("""ViewComponent("foo", model)""")]
+    [DataRow("""ViewComponent(typeof(object))""")]
+    [DataRow("""ViewComponent(typeof(object), model)""")]
+    [DataRow("""Json(model)""")]
+    [DataRow("""Json(model, model)""")]
+    [DataRow("""OnActionExecutionAsync(default(ActionExecutingContext), default(ActionExecutionDelegate))""")]
+    [DataRow("""ViewData""")]
+    [DataRow("""ViewBag""")]
+    [DataRow("""TempData""")]
+    [DataRow("""ViewData["foo"]""")]
+    [DataRow("""ViewBag["foo"]""")]
+    [DataRow("""TempData["foo"]""")]
+    [DataTestMethod]
+    public void ApiControllersShouldNotDeriveDirectlyFromController_DoesNotRaiseWithViewFunctionality(string invocation) =>
+        builder.AddReferences(AspNetCoreReferences)
+            .AddSnippet($$"""
+                using Microsoft.AspNetCore.Mvc;
+
+                [ApiController]
+                public class Invocations : Controller    // Compliant
+                {
+                    object model = null;
+                    public object Foo() => {{invocation}};
+                }
+                """)
+            .VerifyNoIssueReported();
+
+    [DataRow("""OnActionExecuted(default(ActionExecutedContext))""")]
+    [DataRow("""OnActionExecuting(default(ActionExecutingContext))""")]
+    [DataTestMethod]
+    public void ApiControllersShouldNotDeriveDirectlyFromController_DoesNotRaiseWithVoidInvocations(string assignment) =>
+    builder.AddReferences(AspNetCoreReferences)
+        .AddSnippet($$"""
+                using Microsoft.AspNetCore.Mvc;
+                using Microsoft.AspNetCore.Mvc.Filters;
+
+                [ApiController]
+                public class VoidInvocations : Controller           // Compliant
+                {
+                    public void Foo() => {{assignment}};
+                }
+                """)
+        .VerifyNoIssueReported();
+
+    [DataRow("public InConstructor() => foo = View()", DisplayName = "Contrustor")]
+    [DataRow("~InDestructor() => foo = View()", DisplayName = "Destructor")]
```

Comment:
I am confused by the fact that the name of the destructor doesn't match the name of the class in the snippet. I think this should be `~Test() => foo = View()`. Does it compile?

Reply:
No - but why the test is passing is the question....

Reply:
Resolved offline. The `VerifyNoIssue` is not conceived to be used this way. It should be rather used when non-compliant code is under test, but the non-compliances are not raised for some reason (e.g. in a Test context, if the rule is Main).
The correct way of doing it is to use a normal `Verify` with `// Compliant` code (compliant comments are, as usual, not relevant for the verifier, and only defined for better understanding).